dvc/output.py

```diff
@@ -470,6 +484,34 @@ def get_meta(self) -> Meta:
         meta, _ = self._get_hash_meta()
         return meta
 
+    def load_index_meta(self):
+        workspace, key = self.repo.index.get_key(self)
+        index = self.repo.index.data[workspace]
+        entry = index[key]
+        self.hash_info = entry.hash_info
+        self.meta = entry.meta
+        if (
+            not entry.meta
+            and not entry.hash_info
+            or not (
+                (entry.meta and entry.meta.isdir)
+                or (entry.hash_info and entry.hash_info.isdir)
+            )
+        ):
+            return
+
+        self._dir_meta = [
+            {
+                self.PARAM_RELPATH: self.fs.path.join(*subkey[len(key) :]),
+                # order matters here, hash_info takes precedence over any
+                # meta checksums
+                **(entry.meta.to_dict() if entry.meta else {}),
+                **(entry.hash_info.to_dict() if entry.hash_info else {}),
+            }
+            for subkey, entry in index.iteritems(prefix=key)
+            if subkey != key and not (entry.meta and entry.meta.isdir)
+        ]
```

Comment:
Hm, what's the difference between this and a `Tree` object?

Reply:
Since Output's have their own `meta` and `hash_info` which is separate from `out.obj.meta` and `out.obj.hash_info`, then I think dir outputs should have their own dir contents meta as well.

But I was actually wondering if it even makes sense for any (file or dir) output to have it's own hash_info and meta in the first place. It seems like `out.meta` and `out.hash_info` should be properties that just expose `out.obj.meta/out.obj.hash_info`, and `Output.__init__(meta=...)` should always just initialize something like `self.obj = Tree.from_meta()/HashFile.from_meta()`) instead of defaulting to `self.obj = None` and then relying on ODB load later.

The main issue there would be that HashFiles don't have metadata, and the right now the output has to store the file meta separately from the object. It seems a bit strange that Tree objects store the hashfile+meta pairs for each child entry, but for individual files, meta is always separate from the object.

Reply:
@efiop I pushed a change that keeps the tree meta in the `out.obj` Tree and leaves the existing out.hash/out.meta handling alone for now